great_expectations/core/batch_manager.py

```diff
@@ -0,0 +1,164 @@
+from typing import TYPE_CHECKING, Dict, List, Optional
+
+from great_expectations.core.batch import (
+    Batch,
+    BatchData,
+    BatchDefinition,
+    BatchMarkers,
+)
+from great_expectations.core.id_dict import BatchSpec
+
+if TYPE_CHECKING:
+    from great_expectations.execution_engine import ExecutionEngine
+
+logger = logging.getLogger(__name__)
+logging.captureWarnings(True)
+
+
+class BatchManager:
+    def __init__(
+        self,
+        execution_engine: ExecutionEngine,
+        batch_list: Optional[List[Batch]] = None,
+    ) -> None:
+        """
+        Args:
+            execution_engine: The ExecutionEngine to be used to access cache of loaded Batch objects.
+            batch_list: List of Batch objects available from external source (default is None).
+        """
+        self._execution_engine = execution_engine
+
+        self._active_batch_id = None
+        self._active_batch_data_id = None
+
+        if batch_list is None:
+            batch_list = []
+
+        self._batch_cache = OrderedDict()
+        self._batch_data_cache = {}
+
+        self.load_batch_list(batch_list=batch_list)
+        if len(batch_list) > 1:
+            logger.info(
+                f"{len(batch_list)} batches will be added to this Validator.  The batch_identifiers for the active "
+                f"batch are {self.active_batch.batch_definition['batch_identifiers'].items()}"
+            )
+
+    @property
+    def batch_data_cache(self) -> Dict[str, BatchData]:
+        """Dictionary of loaded BatchData objects."""
+        return self._batch_data_cache
+
+    @property
+    def loaded_batch_ids(self) -> List[str]:
+        """IDs of loaded BatchData objects."""
+        return list(self._batch_data_cache.keys())
+
+    @property
+    def active_batch_data_id(self) -> Optional[str]:
+        """The batch id for the default batch data.
+
+        When a specific Batch is unavailable, then the data associated with the active_batch_data_id will be used.
+        """
+        if self._active_batch_data_id is not None:
+            return self._active_batch_data_id
+
+        if len(self._batch_data_cache) == 1:
+            return list(self._batch_data_cache.keys())[0]
+
+        return None
+
+    @property
+    def active_batch_data(self) -> Optional[BatchData]:
+        """The BatchData object from the currently-active Batch object."""
+        if self.active_batch_data_id is None:
+            return None
+
+        return self._batch_data_cache.get(self.active_batch_data_id)
+
+    @property
+    def batch_cache(self) -> Dict[str, Batch]:
+        """Getter for ordered dictionary (cache) of "Batch" objects in use (with batch_id as key)."""
+        return self._batch_cache
+
+    @property
+    def active_batch_id(self) -> Optional[str]:
+        """Getter for active Batch ID"""
+        active_batch_data_id: str = self._active_batch_data_id
+        if active_batch_data_id != self._active_batch_id:
```

Comment:
@billdirks I believe that I had written up an answer in another place in this PR -- in addition, your question is important: yes, `active_batch_data_id` and `active_batch_id` can be different.  The former refers to the most recent loaded `BatchData`, while the latter refers to the most recent requested `Batch`.  There can be situations where no new `BatchData` objects have been loaded; however, a new `Validator` was instantiated and requested `Batch` objects, whose `BatchData` has already been loaded (and cached).  Since `BatchData` IDs are from the same name space as `Batch` IDs, this helps avoid unnecessary loading of data from the backends.  Thanks.